WhosAGoodBoy/Who's A Good Boy/Assets/Scripts/FencedIn.cs

```diff
@@ -0,0 +1,70 @@
+using UnityEngine;
+
+public class FencedIn : MonoBehaviour {
+    private bool _previouslyOutOfBounds;
+
+    public Bounds Fence;
+    public Bounds FencedBound;
+    public bool ShowBounds;
+
+    public delegate void OutOfBoundCallback(Vector3 delta);
+    public event OutOfBoundCallback OnOutOfBounds;
+
+    public delegate void BackInBoundsCallback();
+    public event BackInBoundsCallback OnBackInBounds;
+
+
+    private void Awake()
+    {
+        _previouslyOutOfBounds = false;
+    }
+
+    private void Update()
+    {
+        NotifySubscribers();
+    }
+
+    private void OnDrawGizmos()
+    {
+        if (!ShowBounds) {
+            return;
+        }
+        Gizmos.color = Color.red;
+        Gizmos.DrawWireCube(Fence.center, Fence.size);
+        Gizmos.color = Color.blue;
+        Gizmos.DrawWireCube(FencedBound.center, FencedBound.size);
+    }
+
+    private void NotifySubscribers()
+    {
+        bool outOfBounds = !IsInBounds();
+
+        if (_previouslyOutOfBounds && !outOfBounds) {
+            FireBackInBoundsEvent();
+            _previouslyOutOfBounds = false;
+            return;
+        }
+
```

Comment:
Is this not firing out of bounds events every frame?

Reply:
No, because of the return inside the previous block

Reply:
I tested this and it does indeed fire every frame

Reply:
Oops, so it is. I'll fix it